be/src/exprs/math_functions.cpp

```diff
@@ -899,10 +898,310 @@ StatusOr<ColumnPtr> MathFunctions::cosine_similarity(FunctionContext* context, c
     return result;
 }
 
-// explicitly instaniate template function.
+template <LogicalType TYPE, bool isNorm>
+StatusOr<ColumnPtr> MathFunctions::cosine_similarity2(FunctionContext* context, const Columns& columns) {
+    DCHECK_EQ(columns.size(), 2);
+
+    const Column* base = columns[0].get();
+    const Column* target = columns[1].get();
+
+    if (base->has_null() || target->has_null()) {
+        return Status::InvalidArgument(
+                fmt::format("cosine_similarity does not support null values. {} array has null value.",
+                            base->has_null() ? "base" : "target"));
+    }
+    if (base->is_nullable()) {
+        base = down_cast<const NullableColumn*>(base)->data_column().get();
+    }
+    if (target->is_nullable()) {
+        target = down_cast<const NullableColumn*>(target)->data_column().get();
+    }
+
+    if (!base->is_constant() && !target->is_constant()) {
+        return Status::InvalidArgument(
+                fmt::format("cosine_similarity need a constant array as one of the parameter!"));
+    }
+
+    using CppType = RunTimeCppType<TYPE>;
+    using ColumnType = RunTimeColumnType<TYPE>;
+
+    std::vector<CppType> base_elements;
+
+    auto pattern = ColumnHelper::get_const_value<TYPE_VARCHAR>(base);
+
+    std::string pattern_str = pattern.to_string();
```

Comment:
Could replace `string` with `string_view`, then we needn't create a new `std::string` and the following invocations of `substr` needn't copy anymore.
```suggestion
    std::string_view pattern_str = pattern.to_string();
```